cilium/CFP-TODO-delegated-ipam-cilium-ips.md

```diff
@@ -0,0 +1,117 @@
+
+Unlike other IPAM modes, the cilium-agent daemonset is NOT involved in IPAM. However, several Cilium features require Cilium to assign itself an IP, outside the context of a CNI invocation. These features include endpoint health checking (`endpointHealthChecking.enabled=true`) and ingress controller (`ingressController.enabled=true`). When using delegated IPAM, these features are unavailable and [blocked by validation on cilium-agent startup](https://github.com/cilium/cilium/blob/70ae8d0ef536de807aab849291e5a68758cb8d4d/pkg/option/config.go#L3782).
+
+
+## Goals
+
+* Support endpoint health checking and ingress controller when using Cilium's delegated IPAM mode.
+* The solution should work with any conformant CNI IPAM plugin (avoid assumptions about specifics plugins/platforms).
+* The solution should *not* leak IPs, even if cilium-agent crashes and restarts.
+
+
+## Non-Goals
+
+* This CFP does not propose any changes to other IPAM modes, just to delegated IPAM.
+
+
+## Proposal
+
+### Overview
+
+When it needs to allocate IPs for itself, cilium-agent invokes the delegated IPAM plugin directly.
+
+
+### IPAM Plugin Operations
+
+The delegated IPAM plugin supports these three operations (as of CNI spec 0.4.0):
+
+| Operation  |   Usage                        |      Input                             |   Output                     |
+|------------|--------------------------------|----------------------------------------|------------------------------|
+| ADD        | Allocate an IP                 | CNI_CONTAINERID, CNI_NETNS, CNI_IFNAME | IPs (possibly IPv4 and IPv6) |
+| DEL        | Release an IP                  | CNI_CONTAINERID, CNI_IFNAME            | Success/failure              |
+| CHECK      | Verify that an IP is allocated | CNI_CONTAINERID, CNI_NETNS, CNI_IFNAME | Success/failure              |
+
+(The above table is highly simplified, see the [CNI spec](https://www.cni.dev/docs/spec) for full details.)
+
+The semantics of the above operations differ significantly from how other Cilium IPAM implementations work. In particular, Cilium's `ipam.IPAM` struct supports idempotent allocation of a specific IP using [AllocateIP](https://github.com/cilium/cilium/blob/70ae8d0ef536de807aab849291e5a68758cb8d4d/pkg/ipam/allocator.go#L47). This is used to restore IPs on cilium-agent restart, ensuring that the IP doesn't change and potentially disrupt the dataplane. This isn't possible with delegated IPAM, because:
+
+* The required inputs do not include the IP address. By convention, some [IPAM plugins support an additional "ips" argument](https://www.cni.dev/docs/spec), but this is not universal.
+* The CNI ADD operation is not idempotent. According to [the spec](https://www.cni.dev/docs/spec/#add-add-container-to-network-or-apply-modifications): "A runtime should not call ADD twice (without an intervening DEL) for the same (`CNI_CONTAINERID`, `CNI_IFNAME`) tuple."
```

Comment:
CNI author here. That's why RFCs use "SHOULD" :-p.

But seriously, this is a good motivation to relax the no-add-without-del restriction. We've been idly discussing making ADD idempotent in CNI. I'll take this as another data point.